libtsuba/include/katana/TxnContext.h

```diff
@@ -61,6 +66,8 @@ class KATANA_EXPORT TxnContext {
 
   void SetTopologyWrite() { topology_write_ = true; }
 
+  void DelayManifestCommit() { commit_manifest_ = false; }
```

Comment:
Do we really need to expose this function to callers?

As I understand it, `commit_manifest_` will only be used during the transition from on-disk to in-memory manifests (and even then, I'm not sure if it really is necessary).

Afterwards, people who want to write their manifests (e.g., tests, tools) can just call "Commit(txn_ctx)" (or whatever we call it) themselves.

Reply:
@yige-hu, this is the same issue I'm getting at when I ask about the `commit_manifest_ `boolean.
I'd like to see a CommitToStorage function that is only called by testing code or utilities.  Maybe only `external/katana/tools/graph-convert/`, with a note that this API is deprecated. Or it sounds like we can just remove graph-convert without consequence. 

But the larger issue here is that every Write and Commit should be delayed until there is a commit on the transaction object (however you write it).  That commit will eventually migrate to the server, but needs to be in the code right now.

Reply:
It's not only `external/katana/tools/graph-convert/`, but tons of other code in all sorts of `test` directories under different libraries (e.g. libgraph, libgluon, libquery) that are outside the scope of transactions. Because of all those things not related to operations or transactions (they all calls the functions which requires the `TxnContext` parameter, and we give them a local one created inside the test cases so that there's no segfault under the call stack) which still need to write back the manifest, I'm adding this `bool commit_manifest_`. If you don't like this, we can add something like `TxnContext->CommitToStorage` for every each of the local `TxnContext`-s for the tests/tools that are not related to transactions. But I think it's uglier than setting this boolean with default value true so that it is more transparent to people who don't care about transactions.

Also, it brings back the previous issue of why I ported `AnalyticsOperation` to use `QueryClient`. `TxnManager` maintains its own `QueryClient`. If operations (except for `AnalyticsOperation`, according to previous implementation done to operations) are inside a transaction, they use the `QueryClient` provided by that transaction and commit at the transaction commit. Otherwise, they use the `QueryClient` created by `graph-worker` and commit at the end of each operation. `QueryClient` caches the graph on which it operates. In this way, `TxnManager` can implement a `CommitToStorage` by committing the manifest of the graph cached by `QueryClient`. Otherwise, I'm going to add another pointer in the `TxnManager` which points to the graph that it operates on (due to the current implementation of katana, we only work on one graph every time) and is solely used by `AnalyticsOperation`. Then, we can commit that graph while committing the transaction, which makes it a wrapper as `TxnContext->CommitToStorage`.

Reply:
Also, I don't think we eventually need a `TxnContext->Commit()` API, since ultimately the server is going to write back manifest for operation/transaction related operations. For the tests/tools not related to transactions, I feel it's clearer for the functions to write back manifest as usual, rather than ask the tests/tools to call a commit for each local `TxnContext`.

Reply:
>  It's not only external/katana/tools/graph-convert/, but tons of other code in all sorts of test directories under different libraries (e.g. libgraph, libgluon, libquery) that are outside the scope of transactions. 

My main worry is that the semantics and API of `TxnContext(true)` is subtly different than `TxnContext()`. The first writes to storage if any callee eventually calls `DoStore` while the other doesn't write anything unless you call "Commit(txn_ctx)" and that usually is "delayed" to some top level function.

This raises some questions:
- If one calls "Commit(txn_ctx)" with a context created by `TxnContext(true)` should we commit again? 
- If a caller calls `DelayManifestCommit()` should we commit the manifest so far?
- Do people using the `TxnContext(true)` path depend on commits happening at the time `DoStore` is called? And won't that be different than when we have `TxnContext(false)`? Yet in both cases we have a txn_ctx. How are clients supposed to reason about this difference? 

In my mind it sounds like we have two different ideas.

1. The TxnContext that we know and love and
2. An auto-commit policy for parts of the codebase that don't care too much about transactional execution but would like to call into the part of the codebase that does care

A possible implementation would be:

```cpp
template<typename C>
class CommitGuard {
  C ctx_;
public:
  CommitGuard() {}

  ~CommitGuard() {
    // Yes this is violent but AutoCommit is only for tests
    KATANA_LOG_ASSERT(ctx_.Commit());
  }

  C& operator() {
    return ctx_;
  }
};
```

and in tests and such one would just write
```
AutoCommit<TxnContext> txn_ctx;
```

instead of
```
TxnContext txn_ctx
```

wherever they first needed a transaction context like thing.

If you don't like the idea of separate classes, you could conflate auto commit with TxnContext itself and get something like the boolean constructor you have now and compared to

```
TxnContext(bool auto_commit)
~TxnContext() {
  if (auto_commit_) { ... }
}
```

what you have now is more transparent because the errors would be returned at `DoStore` rather than when the root TxnContext was disposed, though it still looks more magical than it needs to be. And seen from this perspective, an improvement over the current PR would be something like:

```
class TxnContext {
  friend DoStore
  bool auto_commit_;
}
```

because only DoStore really needs to know about the auto commit policy.

Even then, it is a bug for any test to create a context like

```
TxnContext txn_ctx;
```

They must instead write

```
TxnContext txn_ctx(true);
```

which seems more cryptic than

```
AutoCommit<TxnContext> txn_ctx;
```

Reply:
I like the RIAA model of `CommitGuard<TxnContext> txn_ctx;`

I do like this naming better:  `AutoCommit<TxnContext> txn_ctx;`

@yige-hu, I'm not sure I understand your point about `QueryClient `and `TxnManager`, but `QueryClient `doesn't cache the graph anymore.  That is done by `GraphLifetime`.

Reply:
@ddn0 @witchel I can use the `AutoCommit<TxnContext> txn_ctx`. Then I need to cache something to commit in `TxnContext`. Do you want me to cache the `RDGManifest` itself and add/expose another function in `libtsuba`, `PropertyGraph` and `DistPropertyGraph` to write back the `RDGManifest`, or cache a pointer to the `DistPropertyGraph` and calls commit? Can I still assume every `graph-worker` instance works only on one graph, or do I need to keep track of a set of graphs in the future?

Reply:
I think I'm caching the `RDGManifest` in `TxnContext` now. So that all the code remains in `libtsuba`.

Reply:
Feel like we're going to miss this deadline, since I can finish this change as fast as tonight. Then there needs another round of review and changes.... Also, any form of `RDGManifest` committing from the `graph-worker` is eventually not needed, since transaction/operation-related manifests are going to be written back by the server in the near future, and the tests/tools not related to transactions right now still writes back manifest immediately (by default, `TxnContext()` has its `commit_manifest_` set to false, so they only need to define it as `TxnContext()`).

Reply:
@ddn0 @witchel Feel like we're going to miss this deadline, since I can finish this change as fast as tonight. Then there needs another round of review and changes.... Also, any form of RDGManifest committing from the graph-worker is eventually not needed, since transaction/operation-related manifests are going to be written back by the server in the near future, and the tests/tools not related to transactions right now still writes back manifest immediately (by default, TxnContext() has its commit_manifest_ set to false, so they only need to define it as TxnContext()).

Reply:
It makes sense to me that a TxnContext would have an RDGManifest.  That is important metadata to track for a transaction.

I feel like we are talking in circles a bit.  I don't like the `commit_manifest_` flag because of the reasons Donald outlines.  For example, what happens when you call commit on a TxnContext that has `commit_manifest_ == true`?

Yes, the commit will move to the server, but that is M6, not M5.  For now, having an explicit commit operation lets us have confidence that moving the commit operation to the server will work.

Please schedule a zoom call with me if you want to talk about this more.  I'm available tonight and so on.

Reply:
I think I know your idea but can't finish changing today. It makes sense if you want the API in M5 before the server writes back. But is the server writing back manifest not a part of M5?

Reply:
@ddn0 @witchel  I eventually used `bool auto_commit_` in `TxnContext`, and used `TxnContext`'s destructor for the auto commit. I tried the `AutoCommit<TxnContext>` but it doesn't seem to fit in our use cases.
```
template <typename C>
class AutoCommit {
  C ctx_;

public:
  AutoCommit() {}

  ~AutoCommit() {
    // Yes this is violent but AutoCommit is only for tests
    KATANA_LOG_ASSERT(ctx_.Commit());
  }

  operator C&() { return ctx_; }
};
```
With the above code and self-define type converter, it can only convert the following case
```
AutoCommit<TxnContext> a;
TxnContext b = a;
```
But the followed is not possible with C++ self-defined converter, because &a is not a member variable.
```
AutoCommit<TxnContext> a;
TxnContext *b = &a;  // error, not possible
```
  And we need the above case because we call functions in tools and tests as
```
TxnContext txn_ctx;
func(..., &txn_ctx);
```
with AutoCommit, suppose we can convert the pointers,
```
AutoCommit<TxnContext> txn_ctx;
func(..., &txn_ctx);  //error, not possible
```
which needs conversion from `AutoCommit*` to `TxnContext*` if we use AutoCommit with self-defined conversion. Otherwise, the semantics may be ugly and inconvenient for test/tool authors, such as
```
AutoCommit<TxnContext> txn_ctx;
func(..., &((TxnContext)txn_ctx);
```
Or 
```
AutoCommit<TxnContext>  txn_ctx;
func(..., &txn_ctx->GetContext());
```
Therefore, I'd prefer using the `TxnContext` destructor with `bool auto_commit_`.

Reply:
I can. I just feel it looks ugly and inconvenient for people who don't care about transactions (,isn't it?). Similar to the above mentioned 
```
AutoCommit<TxnContext>  txn_ctx;
func(..., &txn_ctx->GetContext());
```